File: studio/core/format.js
// Formatting helpers shared by every screen. No locale/timezone guessing —
// the hub speaks in sim ticks (t) today; wall-clock formatting is added
// when a real node stream carries wall-clock timestamps.

export function fmtNum(v, digits = 1) {
  if (v == null || Number.isNaN(v)) return "—";
  return Number(v).toFixed(digits);
}

export function fmtPct(v, digits = 0) {
  if (v == null || Number.isNaN(v)) return "—";
  return `${Number(v).toFixed(digits)}%`;
}

export function fmtTick(t) {
  return `t=${String(t ?? 0).padStart(3, "0")}`;
}

export function fmtWatts(v) {
  if (v == null || Number.isNaN(v)) return "—";
  return v >= 1000 ? `${(v / 1000).toFixed(2)} kW` : `${Number(v).toFixed(0)} W`;
}

// The five-color health vocabulary (see docs/APP_PLAN.md §1). Screens map
// domain state into one of these tokens; nothing invents its own palette.
export const STATE = Object.freeze({ OK: "st-ok", WARN: "st-warn", CRIT: "st-crit", LEARN: "st-learn", PRED: "st-pred" });

export function fmt(v) {
  if (v == null) return "—";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") return Number.isInteger(v) ? String(v) : v.toFixed(2);
  if (typeof v === "object") return JSON.stringify(v);
  return String(v);
}

export function escapeHtml(s) {
  return String(s).replace(/[&<>]/g, (c) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;" }[c]));
}

File: mobile/lib/core/format.dart
// Formatting helpers shared by every screen — direct Dart port of
// studio/core/format.js. No locale/timezone guessing: the hub speaks in sim
// ticks (t) today; wall-clock formatting is added when a real node stream
// carries wall-clock timestamps.

String fmtNum(double? v, [int digits = 1]) {
  if (v == null || v.isNaN) return '—';
  return v.toStringAsFixed(digits);
}

String fmtPct(double? v, [int digits = 0]) {
  if (v == null || v.isNaN) return '—';
  return '${v.toStringAsFixed(digits)}%';
}

String fmtTick(int? t) => 't=${(t ?? 0).toString().padLeft(3, '0')}';

String fmtWatts(double? v) {
  if (v == null || v.isNaN) return '—';
  return v >= 1000 ? '${(v / 1000).toStringAsFixed(2)} kW' : '${v.toStringAsFixed(0)} W';
}

/// Renders any twin point value — the twin carries bools, numbers, strings
/// and evidence maps through the same {value,t,source,confidence} shape.
String fmtValue(dynamic v) {
  if (v == null) return '—';
  if (v is bool) return v ? 'true' : 'false';
  if (v is num) {
    if (v is int || v == v.roundToDouble()) return v.toInt().toString();
    return v.toStringAsFixed(2);
  }
  if (v is Map || v is List) return v.toString();
  return v.toString();
}
